File: leetcode/leetcode_cpp/nearest-exit-from-entrance-in-maze.cpp
/*
https://leetcode.com/problems/nearest-exit-from-entrance-in-maze

1. BFS
time: o(n)
space: o(n)
*/
class Solution {
public:
    int nearestExit(vector<vector<char>>& maze, vector<int>& entrance) {
        queue<array<int, 3>> q;
        q.push({0, entrance[0], entrance[1]});

        while (!q.empty()){
            auto [step, posX, posY] = q.front(); q.pop();
            if (maze[posX][posY] != '.') continue;
            if (step != 0 && (posX == 0 || posY == 0 || posX == maze.size()-1 || posY == maze[0].size()-1)) return step;
            maze[posX][posY] = 'x';
            
            if (posX+1 < maze.size())
                q.push({step+1, posX + 1, posY});
            if (posX-1 >= 0)
                q.push({step+1, posX - 1, posY});
            if (posY+1 < maze[0].size())
                q.push({step+1, posX, posY+1});
            if (posY-1 >= 0)
                q.push({step+1, posX, posY-1});
        }

        return -1;
    }
};

File: leetcode/leetcode_py/nearest-exit-from-entrance-in-maze.py
'''
https://leetcode.com/problems/nearest-exit-from-entrance-in-maze

1. BFS
time: o(n)
space: o(n)
'''
class Solution:
    def nearestExit(self, maze: List[List[str]], entrance: List[int]) -> int:
        n1 = len(maze)
        n2 = len(maze[0])
        q = [(0, entrance[0], entrance[1])]
        while q:
            step, r, c = q[0]
            q.pop(0)
            if maze[r][c] != '.':
                continue
            if step != 0 and (r == 0 or c == 0 or r == n1 - 1 or c == n2 - 1):
                return step
            maze[r][c] = '+'
            if r + 1 < n1:
                q.append((step+1, r + 1, c))
            if r - 1 >= 0:
                q.append((step+1, r - 1, c))
            if c + 1 < n2:
                q.append((step+1, r, c + 1))
            if c - 1 >= 0:
                q.append((step+1, r, c - 1))
        return -1

        